code/modules/power/engines/tesla/energy_ball.dm

```diff
@@ -104,22 +109,40 @@
 	if(length(orbiting_balls))
 		. += "There are [length(orbiting_balls)] mini-balls orbiting it."
 
-/obj/singularity/energy_ball/proc/move_the_basket_ball(move_amount)
-	var/list/dirs = GLOB.alldirs.Copy()
-	if(length(shocked_things))
-		for(var/i in 1 to 30)
-			var/atom/real_thing = pick(shocked_things)
-			dirs += get_dir(src, real_thing) //Carry some momentum yeah? Just a bit tho
-	for(var/i in 0 to move_amount)
-		var/move_dir = pick(dirs) //ensures teslas don't just sit around
-		if(target && prob(10))
-			move_dir = get_dir(src,target)
-		var/turf/T = get_step(src, move_dir)
+/obj/singularity/energy_ball/proc/move_the_basket_ball(where_to_move)
+	for(var/i in 1 to length(GLOB.field_generator_fields))
+		var/temp_distance = get_dist(src, GLOB.field_generator_fields[i])
+		if(temp_distance <= 15)
+			has_close_field = TRUE
+			break
+	if(has_close_field)
+		var/turf/T = get_step(src, pick(GLOB.alldirs))
 		if(can_move(T))
 			forceMove(T)
-			setDir(move_dir)
+			has_close_field = FALSE
 			for(var/mob/living/carbon/C in loc)
 				dust_mobs(C)
+			return
+		return
+	if(target_turf == null)
+		find_the_basket()
+		return
+	for(var/i in 0 to 8)
+		movement_dir = get_dir(get_turf(src), target_turf)
+		forceMove(get_step(src, movement_dir))
+		if(get_turf(src) == target_turf)
+			target_turf = null
+		for(var/mob/living/carbon/C in loc)
+			dust_mobs(C)
+		has_close_field = FALSE
+
+
+/obj/singularity/energy_ball/proc/find_the_basket()
+	var/area/where_to_move = pick(all_possible_areas) // Grabs a random area that isn't restricted
+	target_area_turfs = get_area_turfs(where_to_move) // Grabs the turfs from said area
+	target_turf = pick(target_area_turfs) // Grabs a single turf from the entire list
+	return
```

Comment:
this variable is only used here, so it can be defined here rather than on the tesla. this avoids holding hard refs to a bunch of turfs.